rust/connlib/tunnel/src/client.rs

```diff
@@ -893,103 +893,97 @@ impl ClientState {
     /// This is important because we don't want to lose state like resolved DNS names for resources that didn't change.
     ///
     /// TODO: Add a test that asserts the above.
-    ///       That is tricky because we need to assert on state deleted by [`ClientState::remove_resources`] and check that it did in fact not get deleted.
+    ///       That is tricky because we need to assert on state deleted by [`ClientState::remove_resource`] and check that it did in fact not get deleted.
     fn set_resources(&mut self, new_resources: Vec<ResourceDescription>) {
-        self.remove_resources(
-            &HashSet::from_iter(self.resource_ids.keys().copied())
-                .difference(&HashSet::<ResourceId>::from_iter(
-                    new_resources.iter().map(|r| r.id()),
-                ))
-                .copied()
-                .collect_vec(),
-        );
+        for id in HashSet::from_iter(self.resource_ids.keys().copied())
+            .difference(&HashSet::<ResourceId>::from_iter(
+                new_resources.iter().map(|r| r.id()),
+            ))
+            .copied()
+        {
+            self.remove_resource(id);
+        }
 
-        self.add_resources(
-            &HashSet::from_iter(new_resources.iter().cloned())
-                .difference(&HashSet::<ResourceDescription>::from_iter(
-                    self.resource_ids.values().cloned(),
-                ))
-                .cloned()
-                .collect_vec(),
-        );
+        for resource in HashSet::from_iter(new_resources.iter().cloned())
+            .difference(&HashSet::<ResourceDescription>::from_iter(
+                self.resource_ids.values().cloned(),
+            ))
+            .cloned()
+        {
+            self.add_resource(resource)
+        }
     }
 
-    pub(crate) fn add_resources(&mut self, resources: &[ResourceDescription]) {
-        for resource_description in resources {
-            if let Some(resource) = self.resource_ids.get(&resource_description.id()) {
-                if resource.has_different_address(resource_description) {
-                    self.remove_resources(&[resource.id()]);
-                }
+    pub(crate) fn add_resource(&mut self, resource: ResourceDescription) {
+        if let Some(resource) = self.resource_ids.get(&resource.id()) {
+            if resource.has_different_address(resource) {
+                self.remove_resource(resource.id());
             }
```

Comment:
Can this ever trigger? Since we only add Resources with different IDs in the above step. Do we test that changing fields on a Resource works?

Reply:
Yes and there was actually a bug here because it is using the same variable name :D